lib/instrumentation/modules/hapi.js

```diff
@@ -14,113 +14,137 @@ module.exports = function (hapi, agent, version) {
 
   debug('shimming hapi.Server.prototype.initialize')
 
-  shimmer.wrap(hapi.Server.prototype, 'initialize', function (orig) {
-    return function () {
-      // Hooks that are always allowed
-      if (typeof this.on === 'function') {
-        this.on('request-error', function (req, err) {
-          agent.captureError(err, {request: req.raw && req.raw.req})
-        })
-
-        this.on('log', function (event, tags) {
-          if (!event || !tags.error) return
-
-          var payload = {custom: event} // TODO: Find better location to put this than custom
-
-          var err = event.data
-          if (!(err instanceof Error) && typeof err !== 'string') {
-            err = 'hapi server emitted a log event tagged error'
-          }
-
-          agent.captureError(err, payload)
-        })
-
-        this.on('request', function (req, event, tags) {
-          if (!event || !tags.error) return
-
-          var payload = {
-            custom: event, // TODO: Find better location to put this than custom
-            request: req.raw && req.raw.req
-          }
-
-          var err = event.data
-          if (!(err instanceof Error) && typeof err !== 'string') {
-            err = 'hapi server emitted a request event tagged error'
-          }
-
-          agent.captureError(err, payload)
-        })
-      } else {
-        debug('unable to enable hapi error tracking')
+  if (semver.satisfies(version, '>=17')) {
+    shimmer.massWrap(hapi, ['Server', 'server'], function (orig) {
+      return function (options) {
+        var res = orig.apply(this, arguments)
+        patchServer(res)
+        return res
       }
-
-      // When the hapi server has no connections we don't make connection
-      // lifecycle hooks
-      if (this.connections.length === 0) {
-        debug('unable to enable hapi instrumentation on connectionless server')
+    })
+  } else {
+    shimmer.wrap(hapi.Server.prototype, 'initialize', function (orig) {
+      return function () {
+        patchServer(this)
         return orig.apply(this, arguments)
       }
+    })
+  }
+
+  function patchServer (server) {
+    // Hooks that are always allowed
+    if (typeof server.on === 'function') {
+      attachEvents(server)
+    } else if (typeof server.events.on === 'function') {
+      attachEvents(server.events)
+    } else {
+      debug('unable to enable hapi error tracking')
+    }
+
+    // When the hapi server has no connections we don't make connection
+    // lifecycle hooks
+    var conns = server.connections
+    if (conns && conns.length === 0) {
```

Comment:
This needs an hapi 17 check since in hapi 17 connections are not a thing any more.

Reply:
The existence check of `server.connections` covers that.

Reply:
I guess what @AdriVanHoudt means is that outputting the debug message "unable to enable hapi instrumentation on connectionless server" isn't really that nice if running hapi 17?

So it should be more like `if (!hapi17plus && conns && cons.length === 0) {` where `hapi17plus` is a boolean that's true only of we're running hapi version 17 or higher.

Reply:
We'd never reach that debug message on hapi 17, because `server.connections` would have to *exist*.

Reply:
Ah my bad 👍

Reply:
it was more for correctness ^^

Reply:
Yeah I see it can be a little confusing to read because of the above comment. How about just updating the comment to something like:

```
// Prior to hapi 17, when the server has no connections we can't make connection
// lifecycle hooks (in hapi 17+ the server always have connections, though the
// `server.connections` property doesn't exists, so this if-statement wont fire)
```

Reply:
updated.